packages/editor/src/constants/GizmoPresets.ts

```diff
@@ -0,0 +1,448 @@
+  material.uniforms.color.value.setHex(colorHex);
+  material.uniforms.visibility.value = visibility;
+  material.opacity = opacity;
+}
+
+// Example materials
+
+
+// Function to get material by name
+export function getGizmoMaterial(name) {
+  const props = materialProperties[name];
+  const material = gizmoMaterialShader.clone()
+  setGizmoMaterialProperties(material, props.color, props.opacity, props.visibility);
+  return material;
+}
+
+export function setGizmoMaterial(material , name) {
+  const props = materialProperties[name];
+  setGizmoMaterialProperties(material, props.color, props.opacity, props.visibility);
+  return material;
+}
+
+// Usage example
+
+
+const matInvisible = getGizmoMaterial(GizmoMaterial.INVISIBLE);
+const matHelper = getGizmoMaterial(GizmoMaterial.HELPER);
+const matRed = getGizmoMaterial(GizmoMaterial.RED);
+const matGreen = getGizmoMaterial(GizmoMaterial.GREEN);
+const matBlue = getGizmoMaterial(GizmoMaterial.BLUE);
+const matRedTransparent = getGizmoMaterial(GizmoMaterial.RED_TRANSPARENT);
+const matGreenTransparent = getGizmoMaterial(GizmoMaterial.GREEN_TRANSPARENT);
+const matBlueTransparent = getGizmoMaterial(GizmoMaterial.BLUE_TRANSPARENT);
+const matWhiteTransparent = getGizmoMaterial(GizmoMaterial.WHITE_TRANSPARENT);
+const matYellowTransparent = getGizmoMaterial(GizmoMaterial.YELLOW_TRANSPARENT);
+const matYellow = getGizmoMaterial(GizmoMaterial.YELLOW);
+const matGray = getGizmoMaterial(GizmoMaterial.GRAY);
+*/
+
+// reusable geometry
+
+const arrowGeometry = new CylinderGeometry(0, 0.04, 0.1, 12).translate(0, 0.05, 0)
+const scaleHandleGeometry = new BoxGeometry(0.08, 0.08, 0.08).translate(0, 0.04, 0)
+const lineGeometry = new BufferGeometry().setAttribute('position', new Float32BufferAttribute([0, 0, 0, 1, 0, 0], 3))
+const lineGeometry2 = new CylinderGeometry(0.0075, 0.0075, 0.5, 3).translate(0, 0.25, 0)
+
+//plane geomerty
+const gizmoPlane = new Mesh(
+  new PlaneGeometry(100000, 100000, 2, 2),
+  new MeshBasicMaterial({
+    visible: false,
+    wireframe: true,
+    side: DoubleSide,
+    transparent: true,
+    opacity: 0.1,
+    toneMapped: false
+  })
+)
+gizmoPlane.layers.set(ObjectLayers.TransformGizmo)
+
+function CircleGeometry(radius, arc) {
+  const geometry = new TorusGeometry(radius, 0.0075, 3, 64, arc * Math.PI * 2)
+  geometry.rotateY(Math.PI / 2)
+  geometry.rotateX(Math.PI / 2)
+  return geometry
+}
+
+// Special geometry for transform helper. If scaled with position vector it spans from [0,0,0] to position
+
+function TranslateHelperGeometry() {
+  const geometry = new BufferGeometry()
+
+  geometry.setAttribute('position', new Float32BufferAttribute([0, 0, 0, 1, 1, 1], 3))
+
+  return geometry
+}
+
+// Creates an Object3D with gizmos described in custom hierarchy definition.
+
+// Gizmo definitions - custom hierarchy definitions for setupGizmo() function
+
+const gizmoTranslate = {
+  X: [
+    [new Mesh(arrowGeometry, matRed), [0.5, 0, 0], [0, 0, -Math.PI / 2]],
```

Comment:
the meshes here still need to be created for each instances of the transform gizmo, otherwise their position/rotation/scale etc will be common between all instances (since we want to support multiple gizmos at a time). It's just the materials/geometries that we want to deduplicate so save on render calls & memory

Reply:
the setup gizmo function clones these

Reply:
Aha okay. It does look like the geometry is cloned. I understand it's to apply a rotation, and I suppose we probably would have to use a custom shader to optimize completely. Let's double check that the material is not cloned, and this should be fine.

Reply:
The materials are not cloned when assigned to the meshes, the meshes are cloned later though